src/ngx-smart-modal/src/services/ngx-smart-modal.service.ts

```diff
@@ -214,4 +207,219 @@ export class NgxSmartModalService {
+      this._dismissModal(e.detail.instance.modal);
+    }) as EventListener);
+
+    window.addEventListener('keyup', this._escapeKeyboardEvent);
+  }
+
+  private _initModal(modalInstance: ModalInstance) {
+    modalInstance.modal.layerPosition += this.getModalStackCount();
+    this.addModal(modalInstance, modalInstance.modal.force);
+
+    if (modalInstance.modal.autostart) {
+      this.open(modalInstance.id);
+    }
+  }
+
+  private _openModal(modal: NgxSmartModalComponent, top?: boolean): boolean {
+    if (modal.visible) {
+      return false;
+    }
+
+    if (top) {
+      modal.layerPosition = this.getHigherIndex();
+    }
+
+    modal.addBodyClass();
+    modal.overlayVisible = true;
+    modal.visible = true;
+    modal.onOpen.emit(modal);
+    modal.markForCheck();
+
+    setTimeout(() => {
+      modal.openedClass = true;
+
+      if (modal.target) {
+        modal.targetPlacement();
+      }
+
+      modal.markForCheck();
+      modal.onOpenFinished.emit(modal);
+    });
+
+    return true;
+  }
+
+  private _toggleModal(modal: NgxSmartModalComponent, top?: boolean): boolean {
+    if (modal.visible) {
+      return this._closeModal(modal);
+    } else {
+      return this._openModal(modal, top);
+    }
+  }
+
+  private _closeModal(modal: NgxSmartModalComponent): boolean {
+    if (!modal.openedClass) {
+      return false;
+    }
+
+    modal.openedClass = false;
+    modal.onClose.emit(modal);
+    modal.onAnyCloseEvent.emit(modal);
+
+    if (this.getOpenedModals().length < 2) {
+      modal.removeBodyClass();
+    }
+
+    setTimeout(() => {
+      modal.visibleChange.emit(modal.visible);
+      modal.visible = false;
+      modal.overlayVisible = false;
+      modal.markForCheck();
+      modal.onCloseFinished.emit(modal);
+      modal.onAnyCloseEventFinished.emit(modal);
+    }, modal.hideDelay);
+
+    return true;
+  }
+
+  private _dismissModal(modal: NgxSmartModalComponent): boolean {
+    if (!modal.openedClass) {
+      return false;
+    }
+
+    modal.openedClass = false;
+    modal.onDismiss.emit(modal);
+    modal.onAnyCloseEvent.emit(modal);
+
+    if (this.getOpenedModals().length < 2) {
+      modal.removeBodyClass();
+    }
+
+    setTimeout(() => {
+      modal.visible = false;
+      modal.visibleChange.emit(modal.visible);
+      modal.overlayVisible = false;
+      modal.markForCheck();
+      modal.onDismissFinished.emit(modal);
+      modal.onAnyCloseEventFinished.emit(modal);
+    }, modal.hideDelay);
+
+    return true;
+  }
+
+  private _deleteModal(modalInstance: ModalInstance) {
+    this.removeModal(modalInstance.id);
+
+    if (!this.getModalStack().length) {
+      modalInstance.modal.removeBodyClass();
+    }
+  }
+
+  /**
+   * Resolve content according to the types
+   * @param content The modal content ( string, templateRef or Component )
+   */
+  private _resolveNgContent<T>(content: Content<T>) {
```

Comment:
I don't understand all, it would be awesome if you could explain me what it's doing 😉